torch_xla/csrc/aten_cpu_fallback.cpp

```diff
@@ -1,34 +1,386 @@
 #include "torch_xla/csrc/runtime/metrics.h"
 #include "torch_xla/csrc/runtime/tf_logging.h"
+#include "torch_xla/csrc/xla_graph_executor.h"
 
 namespace torch_xla {
 
 // TODO(jwtan): Replace this with torch::lazy::Counter. We need
-// _cpu_fallback_counters to remain as torch_xla::runtime::metrics::Counter to
+// _fallback_counters to remain as torch_xla::runtime::metrics::Counter to
 // support torch_xla::runtime::metrics::CreatePerformanceReport(). For more
 // information, see NOTE: [TORCH_LAZY_COUNTER v.s. XLA_COUNTER].
 static std::unordered_map<std::string, ::torch_xla::runtime::metrics::Counter*>
-    _cpu_fallback_counters;
+    _fallback_counters;
 
 // Get all the executed fallback operations.
 // In other words, get all of them whose counters are not zero.
 std::vector<std::string> GetFallbackOperations() {
   std::vector<std::string> fallback;
-  for (auto const& pair : _cpu_fallback_counters) {
+  for (auto const& pair : _fallback_counters) {
     if (pair.second->Value() != 0) {
       fallback.push_back(pair.first);
     }
   }
   return fallback;
 }
 
-void xla_cpu_fallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
+// Most of the functions for the CUDA fallback are a modified version of
+// PyTorch's at::native::cpu_fallback function.
+//
+// Source: aten/src/ATen/native/CPUFallback.cpp
+//
+// While a better solution would be to adapt PyTorch's function to be device
+// agnostic, the changes are not small enough that would make sense for adding
+// just one more device. Therefore, we copied the needed functions in this file.
+//
+// Before each modified function below, we shall specify what has changed,
+// if there was any.
+
+bool UseCUDAFallback() {
+  return runtime::sys_util::GetEnvBool("XLA_FALLBACK_CUDA", false);
+}
+
+struct DeviceInfo {
+  DeviceInfo(c10::Device device, c10::DeviceIndex i = -1)
+      : common_device(device), index(i) {}
+
+  // Synchronizes the CUDA device being used by PyTorch.
+  void synchronize() {
+    TORCH_CHECK(index != -1, "No defined XLA tensors found for CUDA fallback: ",
+                op.operator_name());
+
+    // Save the current PyTorch device, in case it's not the same as the
+    // recorded tensor device.
+    c10::DeviceIndex current = c10::cuda::current_device();
+    c10::cuda::set_device(index);
+    c10::cuda::device_synchronize();
+    c10::cuda::set_device(current);
+  }
+
+  // Common device for all XLA tensors.
+  //
+  // CUDA OpenXLA fallback is supported only when all XLA tensors live in
+  // the same XLA device. This field should be updated and checked every
+  // time we convert an XLA tensor argument into a CUDA tensor.
+  c10::Device common_device;
+
+  // CUDA device index where the tensors live in.
+  //
+  // This is used for synchronizing the device where the fallback operation
+  // was called. This should ensure completion of the CUDA computation, in
+  // order to be used by another XLA computation.
+  c10::DeviceIndex index;
+};
```

Comment:
This struct helps making sure only one device is used in the CUDA OpenXLA fallback.

- XLA tensor devices are checked against `common_device`
- CUDA device `index` is also checked just to be safe